MatrixSDK/ThreePidAdd/MX3PidAddManager.m

```diff
@@ -0,0 +1,706 @@
+                                         userInfo:nil];
+        failure(error);
+        return;
+    }
+
+    MXWeakify(self);
+    threePidAddSession.httpOperation = [self submitValidationToken:token for3PidAddSession:threePidAddSession success:^{
+        MXStrongifyAndReturnIfNil(self);
+
+        MXHTTPOperation *operation;
+        if (self->doesServerSupportSeparateAddAndBind)
+        {
+            // https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928#2b-adding-a-3pid-to-hs-account-after-registration-post-msc2290
+            operation = [self->mxSession.matrixRestClient add3PIDOnlyWithSessionId:threePidAddSession.sid clientSecret:threePidAddSession.clientSecret success:^{
+
+                NSLog(@"[MX3PidAddManager] finaliseAddPhoneNumberSession: DONE: threePid: %@", threePidAddSession);
+
+                threePidAddSession.httpOperation = nil;
+                success();
+
+            } failure:^(NSError *error) {
+                threePidAddSession.httpOperation = nil;
+                failure(error);
+            }];
+        }
+        else
+        {
+            // https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928#2a-adding-a-3pid-to-hs-account-after-registration-pre-msc2290
+            operation = [self->mxSession.matrixRestClient add3PID:threePidAddSession.sid clientSecret:threePidAddSession.clientSecret bind:NO success:^{
+
+                NSLog(@"[MX3PidAddManager] finaliseAddPhoneNumberSession: DONE: threePid: %@", threePidAddSession);
+
+                threePidAddSession.httpOperation = nil;
+                success();
+
+            } failure:^(NSError *error) {
+                threePidAddSession.httpOperation = nil;
+                failure(error);
+            }];
+        }
+
+        if (operation)
+        {
+            [threePidAddSession.httpOperation mutateTo:operation];
+        }
+
+    } failure:^(NSError *error) {
+        threePidAddSession.httpOperation = nil;
+        failure(error);
+    }];
+}
+
+
+#pragma mark - Bind Email
+
+- (MX3PidAddSession*)startIdentityServerEmailSessionWithEmail:(NSString*)email
+                                                         bind:(BOOL)bind
+                                                      success:(void (^)(BOOL needValidation))success
+                                                      failure:(void (^)(NSError * _Nonnull))failure
+{
+    MX3PidAddSession *threePidAddSession = [[MX3PidAddSession alloc] initWithMedium:kMX3PIDMediumEmail andAddress:email];
+    threePidAddSession.bind = bind;
+
+    NSLog(@"[MX3PidAddManager] startIdentityServerEmailSessionWithEmail (bind:%@) : threePid: %@", @(bind), threePidAddSession);
+
+    [self startIdentityServer3PidSession:threePidAddSession success:success failure:failure];
+
+    return threePidAddSession;
+}
+
+- (void)tryFinaliseIdentityServerEmailSession:(MX3PidAddSession*)threePidAddSession
```

Comment:
Maybe rename `tryFinaliseIdentityServerBindEmailSession to add more context like `tryFinaliseAddEmailSession` above.